### PR TITLE
WER metric with tests

### PR DESCRIPTION
**Pull request by okuchaiev**
This PR implements WER computation using PTL.metrics.TensorMetric approach.
It adds it to ASR model for both training and evaluation.
Also adds unittests for WER

@vsl9  and @redoctopus  could you please double-check that WER computation logic is actually correct?

### CODE REVIEW

**Comment by okuchaiev** [comment]
> Generally, the PR looks good to me.
> However, name "AverageTextWER" might be vague: here we compute overall WER and not average across utterances WER. "Overall WER" looks like a more commonly accepted term.

- Should I call it just "WER" or "OverallWER" ?

**Comment by okuchaiev** [comment]
@titu1994  just FYI for WER metric

File: nemo/collections/asr/metrics/wer.py [hunk 1]
```diff
@@ -52,88 +49,105 @@ def word_error_rate(hypotheses: List[str], references: List[str], use_cer=False)
...
-      eval_metric: An optional string from 'WER', 'CER'. Defaults to 'WER'.
+        vocabulary:
+        batch_dim_index:
+        use_cer:
+        ctc_decode:
+
     Returns:
-      batch wer, hypothesis and reference from the first batch element
+        res: a torch.Tensor object with two elements: [wer_nominator, wer_denominator]. To correctly compute average
+        text word error rate, compute wer=wer_nominator/wer_denominator
```
Comment (redoctopus) [inline nemo/collections/asr/metrics/wer.py:91]: Nitpick: `s/nominator/numerator/` and `s/wer_num, wer_denum/wer_num, wer_denom/`

File: nemo/collections/asr/metrics/wer.py [hunk 2]
```diff
@@ -52,88 +49,105 @@ def word_error_rate(hypotheses: List[str], references: List[str], use_cer=False)
...
-    Takes output of greedy ctc decoder and performs ctc decoding algorithm to
-    remove duplicates and special symbol. Prints sample to screen, computes
-    and logs AVG WER to console
+    This metric computes nominator and denominator for Average Text Word Error Rate.
+    When doing distributed training/evaluation the result of res=AverageTextWER(predictions, targets, target_lengths) calls
+    will be all-reduced between all workers using SUM operations.
+    Here contains two numbers res=[wer_nominator, wer_denominator]
+
+    If used with PytorchLightning LightningModule, include wer_nominator and wer_denominator inside validation_step results.
+    Then aggregate (sum) then at the end of validation epoch to correctly compute validation WER.
```
Comment (redoctopus) [inline nemo/collections/asr/metrics/wer.py:68]: Should we also make it explicit in the docstring that to get WER one would need to do `WER=res[0]/res[1]` after summing both?

File: nemo/collections/asr/models/ctc_models.py [hunk 3]
```diff
@@ -173,8 +175,8 @@ def training_step(self, batch, batch_nb):
         loss_value = self.loss(
             log_probs=log_probs, targets=transcript, input_lengths=encoded_len, target_lengths=transcript_len
         )
-        wer = self.__wer(predictions, transcript, transcript_len)
-        tensorboard_logs = {'train_loss': loss_value, 'training_wer': wer}
+        wer_num, wer_denum = self.__wer(predictions, transcript, transcript_len)
```
Comment (redoctopus) [inline nemo/collections/asr/models/ctc_models.py:178]: Same nitpick as before (denum -> denom), here and in the following changed lines

**Comment by vsl9** [comment]
> Should I call it just "WER" or "OverallWER" ?
"WER" looks more favorable to me. "Overall WER" is the term that could be found in some papers for this overall (total, global) WER when we accumulate number of errors for all utterances and then divide it by number of words. "Average" might be confused with averaging WERs that were computed for each utterance individually [which is slightly different metric].

**Comment by okuchaiev** [comment]
addressed feedback

**Pull request merged**